tmt/utils.py

```diff
@@ -7270,3 +7271,80 @@ def render_rst(text: str, logger: Logger) -> str:
     document.walkabout(visitor)
 
     return visitor.rendered
+
+
+def jira_link(
+        nodes: list[Union['tmt.base.Test', 'tmt.base.Plan', 'tmt.base.Story']],
+        links: 'tmt.base.Links',
+        separate: Optional[bool] = False) -> None:
+    """ Link the object to Jira issue and create the URL to tmt web service """
+
+    def create_url(tmt_object: 'tmt.base.Core') -> list[str]:
+        url: list[str] = []
+        # Get the fmf id of the object
+        if isinstance(tmt_object, tmt.base.Test):
+            fmfid = tmt_object.fmf_id
+            tmt_type = "test"
+        elif isinstance(tmt_object, tmt.base.Plan):
+            fmfid = tmt_object.fmf_id
+            tmt_type = "plan"
+        elif isinstance(tmt_object, tmt.base.Story):
+            # TODO: not supported by the service yet
+            fmfid = tmt_object.fmf_id
+            tmt_type = "story"
+        url.append(f'{tmt_type}-url={fmfid.url}')
+        url.append(f'{tmt_type}-name={fmfid.name}')
+        if fmfid.path is not None:
+            url.append(f'{tmt_type}-path={fmfid.path}')
+        if fmfid.ref is not None:
+            url.append(f'{tmt_type}-ref={fmfid.ref}')
+        return url
+
+    def construct_url_from_list(url: str, url_parts: list[str]) -> str:
+        joined_url = '&'.join(url_parts)
+        joined_url = url + joined_url
+        # Ask for HTML format (WIP, can be changed once FE web server is implemented)
+        joined_url += '&format=html'
+        return joined_url
+
+    # Setup config tree
+    config_tree = tmt.utils.Config()
+    print(config_tree.fmf_tree.find('/user/linking').data)
+    # Linking is not setup in config, therefore user does not want to use linking
```

Comment:
How do you know? :) The user just called `tmt link`, that seems to me like a solid argument that the user wanted to use linking :) Or the user called `test create --link ...`, then I guess the user is interested in linking... It's hard to tell from here, but most likely the opposite is true - `jira_link` would not get called if somebody didn't try to link things to Jira.

My two cents: the function does too many things. It takes care of links itself, but also prepares its own inputs, reads configuration, tries to look for a `linking` stanza in it, then picks a single link out of `links` - `verifies` only, no other relation is supported. It has so many inputs. I would recommend splitting it in two:

* a function that focuses on creating links and nothing else. It's given a linking configuration, a single `Link` instance, and a list of nodes. And the `separate`, IIUIC. It has a simple task, talk to Jira, construct a linking service URL, and submit changes to Jira. No looking around, no external state to load or test, nothing to decide on its own, input-driven.
* a function that manages these links, some kind of "ticket linking manager", something that's responsible for adding links to (various) services like Jira or Bugzilla or Trello, whatever. Give it nodes, give it a list of `Link` instances. It loads the configuration, finds what entries are there for `linking`, call the respective linking function for them and the given links - `type: jira` => `jira_link()`, raise an error for unknown `type` in the `linking` entry, unsupported. It may return the list of successfully processed links and leave it up to the CLI and `base.py` to decide whether it's fine or not that no Jira link was actually created, or it may raise an exception - or re-raise an exception raised by `jira_link()` if the linking fails.

I understand it's a WiP and a prototype, yet I'd recommend focusing on the separation of concerns. Otherwise, it will become harder and harder to extend and test.

Reply:
As I think the separation to two functions is a good idea, I am not sure if the second function you mentioned should take a list of `Link` instances as a argument. AFAIK, currently, the idea was that we link in a one-to-many relation, several tmt objects can be linked to one issue. Also, supporting more types other than Jira is out of scope of this PR as well, but I agree that it is a good idea to refactor the code now to make it ready for adding more types in the future easily.

Reply:
The approach outlined by @happz seems reasonable to me. Regarding the question around providing a list of `Link` instances: Currently the command line syntax for `tmt link` takes only a single link as input:

    tmt link verifies:https://foo.com/a/b/c /plans/core /tests/smoke

It is true that `tmt {test|plan|story} create` commands can take multiple `--link` options so there seems to be use case also for multiple links. Thinking about it, should we change the design of `tmt link` to allow multiple links as well? Or are we ok with the currently proposed way?

Reply:
> How do you know? :) The user just called `tmt link`, that seems to me like a solid argument that the user wanted to use linking :) Or the user called `test create --link ...`, then I guess the user is interested in linking...

There are actually two parts of the linking action:

* add the link to the test, plan or story metadata, this is now done by `tmt * create` only
* link the web service url to the issue tracking service

It's perfectly fine if user does not have issues trackers defined in their config to only perform the first step. So calling `tmt link` with no config would also make sense for example for quickly updating metadata on disk for several tests at once.

Implementation of the second step should not be to much difficult as `fmf` already supports a comfortable way to modify existing metadata on disk using a `with` statement. Here's an example for inspiration:

https://github.com/teemtee/tmt/blob/e6a0964882019f39eb3a62eade4171ef2c21e5ca/tmt/export/nitrate.py#L438-L439

Reply:
> It's perfectly fine if user does not have issues trackers defined in their config to only perform the first step. So calling `tmt link` with no config would also make sense for example for quickly updating metadata on disk for several tests at once.

It can be a bit tricky - 'extra-nitrate' has to be in leaf node. But some links can be already inherited/merged/adjusted - so in many cases it will modify the fmf files more than user would assume (eg. set link to the processed value")

```

```

Reply:
As suggested above, in 7ecd3bb7 I've separated code into more functions to make it more readable and maintainable. I believe it shoud cover the core points.